### ISSUE TITLE
Focus follows mouse: proposed behavior change

### ISSUE THREAD

**Issue by jackric** [koekeishiya/kwm#476]
This situation frequently annoys me: I'm in floating mode, with a large dominant window (e.g. Chrome), with some other windows behind it. As I move the mouse up to the menubar, one of the background windows catches the focus because it had a bit of area exposed (see screenshot).

I have 3 ideas to make kwm less annoying in this situation:
1. An option to disable `focus-follows-mouse` when **a)** tiling mode is _floating_ and **b)** a dominant window has focus. We can classify a dominant window as one having a high percentage of screen space.
2. An option to disable `focus-follows-mouse` when tiling mode is _floating_.
3. A configurable delay for capturing focus, e.g. cursor must be in window space for 10ms before focus is given to that window. My fast-moving cursor travelling to the menubar therefore won't trigger other windows.

I think (1) may be difficult to form "muscle memory" around, (2) may be annoying when I have carefully-placed floating windows (e.g. GIMP) and (3) may introduce a perceptible delay in all other workflows.

I'm willing to work on a pull request, but keen to gather thoughts on these approaches. @koekeishiya ? 
![image](https://cloud.githubusercontent.com/assets/259713/18461093/ebb02b5e-79b9-11e6-93d6-ac7d117aebb1.png)

**Comment by koekeishiya** [comment]
From my pov, I find that all of these brnig theur own advantages and disadvantages, and I don't really find any of them to be objectively good. If anything, having a config setting for option 2 would probably be the approach I'd choose.

I still think that simply having a keybind to toggle focus-follow-mouse is the appropriate solution.

You are of course free to modify the source as you'd like, but I don't think it is something that really belongs in the upstream version.

**Comment by chrishoage** [comment]
Just pasting info here, kwm has an option to disable FFM when a floating window has focus:

```
// Disable autorise when a floating window is focused
kwmc config standby-on-float on
```

Which sounds like it does half of what you're requesting.

**Comment by thonixx** [comment]
Let me refer to the ffm implementation in Linux (Compiz). They have several settings including en-/disabling focus follows mouse, en-/disabling auto-raise (just to be sure we are on the same page - auto-raise works that the focused window will also be raised to the foreground. in some implementations of ffm the window just gets focus but does not get raised to the top and stays. I am not sure if this is even possible on OSX.) and setting delays for the auto-raise so a window gets a focus but doesn't raise for 500ms.

This was my preferred setting on Linux (Compiz) and it worked fine by having a floating mode desktop with focus-follows-mouse and auto-raise as it was possible to move the mouse to other areas without having windows auto-raised which I don't really need.

Currently I have kwm with floating mode enabled and with having a big terminal window and some small windows (i.e. a VoIP client) in the background. So I go to the docker, click on the VoIP client and it appears in the middle of the terminal window. Then I like to use the application and immediately the terminal window gets auto-raised again and the VoIP window disappears. So I have to switch with CMD-<Tab> and hope that I am in the right location as the terminal window gets focus again when I switch with CMD-<Tab> and when I am not in the right position with my cursor.

Having some sort of auto-raise delay would solve many issues with ffm/kwm with my workflow. And it would not hurt to have a setting of the delay as people can set it like they want. So if one likes to have the current behaviour they set the delay to 0ms and if one wants to "disable" the delay they set something like 20 seconds so it doesn't auto-raise for a long period. Just a thought from my side. And the most annoying thing in kwm - otherwise kwm is great but I had to switch to floating mode for some reasons.

**Comment by koekeishiya** [comment]
> Please let me refer to the ffm implementation in Linux (Compiz). They have several settings including en-/disabling focus follows mouse, en-/disabling auto-raise (just to be sure we are on the same page - auto-raise works that the focused window will also be raised to the foreground. in some implementations of ffm the window just gets focus but does not get raised to the top and stays. I am not sure if this is even possible on OSX.) and setting delays for the auto-raise so a window gets a focus but doesn't raise for 500ms.

I completely agree with this, however, the problem is:

> the window just gets focus but does not get raised to the top and stays. I am not sure if this is even possible on OSX.)

This is not possible to do in OSX. It is either autoraise or no focus-follows-mouse at all.
That is why I think the delay is kinda pointless, as it doesn't only delay the "raise". It also
prevents the actual change in focus.

**Comment by thonixx** [comment]
Let me rephrase:

Please introduce a delay in focus-follows-mouse until a window gets focused and auto-raised at the same time as it is not possible to have a separate auto-raise delay.

I made some research and stumbled upon the following Stackoverflow discussion:
http://stackoverflow.com/a/35073462

This solves the problem as it has a configurable delay in 50ms steps. So I have it running with 250ms delay and it works like a charm but it would be nice to have this behaviour in kwm.

**Comment by akdor1154** [comment]
I have focus follows mouse disabled for the single reason that it can make it very difficult to get to the menu bar. A delay would help alleviate this, though there may be other approaches too.

**Comment by thonixx** [comment]
No further input here? I switched to the AutoRaise script I have posted above and I am happy with it. Maybe it gets implemented in kwm to use the tiling function with a better ffm principle again.

**Comment by koekeishiya** [comment]
I still do not think simply a delay-setting is a good solution.
Will post back as I figure out some proper solution for this.

**Comment by koekeishiya** [comment]
I personally have so far found the best solution to simply prevent focus-follows-mouse from triggering while say, the `alt-key` is down. This is the approach I've taken in **chunkwm**, and I find this to be better than an artificial delay.

**Comment by Amar1729** [comment]
Will this be the default behavior in chunkwm? And is something like this possible in kwm right now while chunkwm is still in development?

**Comment by koekeishiya** [comment]
> Will this be the default behavior in chunkwm? And is something like this possible in kwm right now while chunkwm is still in development?

Yes and yes.

**Comment by dominiklohmann** [comment]
Shouldn't releasing the alt-key now trigger focus-follows-mouse? Right now you need to move the cursor again to focus the window under it after releasing the alt-key.
